flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZookeeperLeaderRetrievalDriver.java

```diff
@@ -0,0 +1,183 @@
+	private static final Logger LOG = LoggerFactory.getLogger(ZookeeperLeaderRetrievalDriver.class);
+
+	/** Connection to the used ZooKeeper quorum. */
+	private final CuratorFramework client;
+
+	/** Curator recipe to watch changes of a specific ZooKeeper node. */
+	private final NodeCache cache;
+
+	private final String retrievalPath;
+
+	private volatile boolean running;
+
+	private final ConnectionStateListener connectionStateListener = (client, newState) -> handleStateChange(newState);
+
+	private DefaultLeaderRetrievalService.StateHandler leaderRetrievalStateHandler;
+
+	/**
+	 * Creates a leader retrieval service which uses ZooKeeper to retrieve the leader information.
+	 *
+	 * @param client Client which constitutes the connection to the ZooKeeper quorum
+	 * @param retrievalPath Path of the ZooKeeper node which contains the leader information
+	 */
+	public ZookeeperLeaderRetrievalDriver(CuratorFramework client, String retrievalPath) {
+		this.client = checkNotNull(client, "CuratorFramework client");
+		this.cache = new NodeCache(client, retrievalPath);
+		this.retrievalPath = checkNotNull(retrievalPath);
+
+		running = false;
+	}
+
+	@Override
+	public void start(DefaultLeaderRetrievalService.StateHandler stateHandler) throws Exception {
+		LOG.info("Starting {}.", this);
+
+		this.leaderRetrievalStateHandler = checkNotNull(stateHandler);
+
+		client.getUnhandledErrorListenable().addListener(this);
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(connectionStateListener);
+
+		running = true;
+	}
+
+	@Override
+	public void stop() throws Exception {
+		if (!running) {
+			return;
+		}
+
+		running = false;
+
+		LOG.info("Stopping {}.", this);
+
+		client.getUnhandledErrorListenable().removeListener(this);
+		client.getConnectionStateListenable().removeListener(connectionStateListener);
+
+		try {
+			cache.close();
+		} catch (IOException e) {
+			throw new Exception("Could not properly stop the ZooKeeperLeaderRetrievalService.", e);
+		}
+	}
+
+	@Override
+	public void nodeChanged() {
+		if (running) {
+			leaderRetrievalStateHandler.notifyIfNewLeaderAddress(() -> {
+				LOG.debug("Leader node has changed.");
+
+				final ChildData childData = cache.getCurrentData();
+
+				final String leaderAddress;
+				final UUID leaderSessionID;
+
+				if (childData == null) {
+					leaderAddress = null;
+					leaderSessionID = null;
+				} else {
+					byte[] data = childData.getData();
+
+					if (data == null || data.length == 0) {
+						leaderAddress = null;
+						leaderSessionID = null;
+					} else {
+						ByteArrayInputStream bais = new ByteArrayInputStream(data);
+						ObjectInputStream ois = new ObjectInputStream(bais);
+
+						leaderAddress = ois.readUTF();
+						leaderSessionID = (UUID) ois.readObject();
+					}
+				}
+				return new Tuple2<>(leaderSessionID, leaderAddress);
```

Comment:
I think the whole address extraction does not have to happen under the lock. I'd suggest to do it first and then simply call `LeaderRetrievalEventHandler.notifyLeaderAddress`.